include/rwqueue.h

```diff
@@ -134,37 +132,52 @@ class RWQueue {
-	// return (and can be reused). The method potentially blocks until there
-	// is enough capacity in the queue for the new items.
+	// Items are std::move'd out of the source vector and also erased so the
+	// source vector is returned in a defined state. The method potentially
+	// blocks until there is enough capacity in the queue for the source items.
 
 	// If queuing has stopped prior to bulk enqueing, then this will
-	// immediately return false and no items will be queued.
+	// immediately return with 0 and no items will be queued.
 
 	// If queuing stops in the middle of enqueing prior to completion, then
-	// this will immediately return false. The items queued /before/
-	// stopping will be available in the queue however the items that came
-	// after stopping will not be queued.
-	bool BulkEnqueue(std::vector<T>& from_source, const size_t num_requested);
+	// this will immediately return the quantity dequeued. The items queued
+	// /before/ stopping will be available in the queue however the items
+	// that came after stopping will not be queued and will still be in the
+	// source vector.
+	size_t BulkEnqueue(std::vector<T>& from_source, const size_t num_requested);
+
+	// Bulk enqueue all of the source vector's items
+	size_t BulkEnqueue(std::vector<T>& from_source);
 
 	// Does nothing if queue is at capacity or queue is not running.
-	// Otherwise, enqueues as many elements as possible until the queue is at capacity.
-	// Moved elements will be erased from the source vector.
+	// Otherwise, enqueues as many elements as possible until the queue is
+	// at capacity. Moved elements will be erased from the source vector.
 	// Elements not enqueued will be left in the source vector.
 	// Returns the number of elements enqueued.
 	size_t NonblockingBulkEnqueue(std::vector<T>& from_source, const size_t num_requested);
 
-	// The target vector will be resized to accomodate, if needed. The
-	// method potentially blocks until the requested number of items have
-	// been dequeued.
+	// Bulk enqueue all of the source vector's items
+	size_t NonblockingBulkEnqueue(std::vector<T>& from_source);
 
-	// If queuing has stopped:
-	//  - Returns true when  one or more item(s) have been dequeued.
-	//  - Returns false when no items can be dequeued.
+	// Deque's the requested number of items into the given container
+	// in-bulk, and returns the quantity dequeued. This potentially blocks
+	// until the requested number of items have been dequeued. Normally all
+	// of the requested items are dequeued, however:
+	//
+	// - If queuing was stopped prior to bulk dequeueing then this
+	//   immediately returns with a value of 0 and no items dequeued.
 	//
-	// The vector is always sized to match the number of items returned.
-	bool BulkDequeue(std::vector<T>& into_target, const size_t num_requested);
+	// - If queuing was stopped in the middle of bulk dequeueing then this
+	//   immediately returns with a value indicating the subset dequeued.
```

Comment:
Is there any value in providing API guarantees for "stop queue while performing an action" behaviour? I don't think so as enqueing normally you only stop the queue on shutdown or subsystem reinit' right?

Would it simplify the code, the contract, and the doco to just stipulate at the top of the class in a comment to ensure no enque/dequeue should be happening during shutting down if you care about well-specified, deterministic behaviour? I think it's reasonable to push the responsibility of waiting before shutdown until all enque/dequeue ops have finished to the client, otherwise undefined behaviour is fine.

I'd wager we don't even care about proper queue shutdown that during reinit, right? And surely not during the final shutdown on emulator exit.

Reply:
The queues do get stopped at runtime inside the `MIXER_LockMixerThread()` function.  Otherwise you get deadlocks.  Most common situation this happens is when changing mixer settings.

It doesn't simplify anything to change this.  Logic looks like this (in pseudo code):

```cpp
BulkDequeue()
{
    while(queue_is_running && output.size() < requested_size) {
        cond_var.wait_until(has_more_data || queue_is_stopped);
        output_data.enqueue(audio_data);
    }
}
```

When the queue gets stopped, it returns immediately instead of waiting for more data.  The main point of stopping the queue is to make the consumer stop waiting for more data so that another thread can acquire the global mixer mutex.

Reply:
Yes; even if the bookkeeping granularity was made coarser - and we only traked bulk actions at the "chunk" level (and ignored individual item quantities), the "issue" is that the while loop is already performing enmasse `std::move`s of swaths of items (with a single algorithm call), in largest chunk size possible.

I also checked if we could move the notification outside the bulk while loops - but that's also impossible because it can lead to deadlocks (which makes sense: if we're enqueuing or dequeuing more items in bulk than the  queue's size.. those notifications after each chunk will "pump" the other end of the queue to relieve the back or front pressure).

Reply:
Cool, I was just wondering 😄